### ISSUE TITLE
perf(qwen3): make fused QKV and gate/up GEMMs parity-safe

### ISSUE THREAD

**Issue by xiaguan** [pegainfer-project/pegainfer#746]
## Context

Qwen3 already stores related projection weights contiguously at load time:

- `qkv_proj = vstack(q_proj, k_proj, v_proj)`
- `gate_up_proj = vstack(gate_proj, up_proj)`

The forward path still reads row ranges from those matrices and launches separate GEMMs: three for Q/K/V and two for gate/up. `openinfer-qwen3/src/batch_decode.rs` records why the previous fused-QKV path was not retained: under shard-local tensor parallelism, its numerical differences were large enough to flip greedy tokens in parity tests.

This is a focused follow-up to #456, which identifies fused QKV and gate/up projection as known roofline levers.

## Goal

Make the contiguous weight layout useful for compute by providing parity-safe fused projection paths:

1. one GEMM for Q/K/V followed by a lightweight split/deinterleave step;
2. one GEMM for gate/up followed by the existing SwiGLU elementwise step.

The implementation must preserve the current eager and CUDA Graph execution contracts and must fail closed for unsupported TP shapes.

## Scope

- Cover Qwen3 TP=1 and TP=2 first.
- Diagnose the current fused-QKV numerical divergence before changing the production path.
- Reuse the existing contiguous `qkv_proj` and `gate_up_proj` allocations.
- Keep LoRA behavior correct for logical q/k/v and gate/up projections.
- Avoid adding a generic abstraction unless another model has the same immediate need.

### Correctness

- Existing Qwen3 HF golden gates pass for TP=1.
- TP=2 greedy/logits parity passes on the current tensor-parallel coverage.
- Eager and CUDA Graph paths produce equivalent results under the repository's existing tolerances.
- LoRA-enabled and LoRA-disabled paths remain covered.

### Performance

Report measured A/B data before claiming a win:

- current split GEMMs vs fused projection path;
- TP=1 and TP=2;
- prefill and decode;
- at least concurrency 1 and 8;
- kernel timings plus end-to-end TPOT/throughput.

Keep the fused path only where the measured end-to-end result improves without weakening correctness gates.

## Relevant code

- `openinfer-qwen3/src/weights.rs`: load-time `DeviceMatrix::vstack`
- `openinfer-qwen3/src/batch_decode.rs`: split Q/K/V and gate/up GEMMs; parity note
- `openinfer-qwen3/src/prefill.rs`: row-range projection GEMMs
- `openinfer-qwen3/src/unified_forward.rs`: row-range projection GEMMs

Related: #456

**Comment by Polaris3003** [comment]
Hi, I'd like to explore this. Could you assign this to me?

**Comment by Polaris3003** [comment]
## Qwen3 fused projection validation results

I completed the Qwen3-4B fused-projection validation on a 2-GPU SM86 machine.

This comment reports the collected correctness, numerical, topology, and E2E
performance data for review. It does not make an Auto-whitelist recommendation.

### Test environment

- GPU architecture: `sm_86`, 2 GPUs
- CUDA toolkit: 12.6
- Model: Qwen3-4B
- Numeric policy: `Tuned`
- Decode overlap: off
- TP sizes: 1 and 2
- E2E concurrency: 1 and 8
- Two repetitions per concurrency
- Model hash: `4b5fcd3463ccfb909a9cf41e5ccabfae1ee4310eb022077434932b0e1f42f0e1`

### Validation coverage

| Validation area | Result |
| --- | --- |
| Correctness matrix | **PASS, 20/20** |
| Projection numerical reports | **PASS, 3/3** |
| Decode topology matrix | **PASS, 16/16** |
| E2E benchmark matrix | **complete, 64/64** |
| Benchmark command errors | **0** |

The projection reports cover:

- TP1 rank 0
- TP2 rank 0
- TP2 rank 1

The correctness matrix covers split, QKV-only, gate/up-only, and both-fused
modes under TP1 and TP2, including HF and LoRA golden gates. Existing
correctness tolerances were not changed.

### Measurement definition

For the tables below:

- a positive latency improvement means the fused candidate had lower latency;
- a negative latency improvement means the fused candidate had higher latency;
- prefill/unified latency uses p50 TTFT;
- decode latency uses p50 steady-state TPOT;
- a positive throughput improvement means the fused candidate produced higher
  throughput;
- each concurrency cell is shown as `repeat 1 / repeat 2`.

The suite's reference thresholds are:

- prefill/unified: 3%
- decode: 2%

These thresholds are included for context; no whitelist decision is made in
this report.

### Latency improvement

| Projection | Phase | TP | Concurrency 1 (R1 / R2) | Concurrency 8 (R1 / R2) | Four-run mean | Reference threshold |
| --- | --- | ---: | ---: | ---: | ---: | ---: |
| QKV | prefill/unified | 1 | +1.30% / +1.44% | +2.42% / +4.44% | **+2.40%** | 3.0% |
| gate/up | prefill/unified | 1 | +0.28% / -0.56% | +2.53% / +2.95% | **+1.30%** | 3.0% |
| QKV | decode | 1 | +2.59% / +2.79% | +1.37% / +2.51% | **+2.32%** | 2.0% |
| gate/up | decode | 1 | +0.45% / +1.25% | +0.92% / -0.16% | **+0.61%** | 2.0% |
| QKV | prefill/unified | 2 | -3.33% / +4.30% | -4.46% / +3.01% | **-0.12%** | 3.0% |
| gate/up | prefill/unified | 2 | -4.53% / +1.40% | -4.31% / -1.10% | **-2.13%** | 3.0% |
| QKV | decode | 2 | -2.98% / +11.21% | +46.13% / -7.60% | **+11.69%** | 2.0% |
| gate/up | decode | 2 | -2.41% / +2.63% | -6.94% / -5.11% | **-2.96%** | 2.0% |

### Throughput improvement

| Projection | Phase | TP | Concurrency 1 (R1 / R2) | Concurrency 8 (R1 / R2) | Four-run mean |
| --- | --- | ---: | ---: | ---: | ---: |
| QKV | prefill/unified | 1 | +0.95% / +1.16% | +2.83% / +4.20% | **+2.28%** |
| gate/up | prefill/unified | 1 | +0.61% / -0.77% | +1.89% / +3.18% | **+1.23%** |
| QKV | decode | 1 | +2.53% / +2.93% | +1.45% / +2.61% | **+2.38%** |
| gate/up | decode | 1 | +0.13% / +1.32% | +0.87% / -0.21% | **+0.53%** |
| QKV | prefill/unified | 2 | -3.57% / +4.57% | -4.65% / +3.16% | **-0.12%** |
| gate/up | prefill/unified | 2 | -4.58% / +1.17% | -4.27% / -1.03% | **-2.18%** |
| QKV | decode | 2 | -2.86% / +12.71% | +85.02% / -7.03% | **+21.96%** |
| gate/up | decode | 2 | -2.49% / +2.68% | -6.58% / -4.78% | **-2.79%** |

### Repeat consistency and kernel direction

| Projection | Phase | TP | All latency repeats positive | No large latency regression | Throughput consistent | Kernel aggregate follows E2E direction |
| --- | --- | ---: | --- | --- | --- | --- |
| QKV | prefill/unified | 1 | Yes | Yes | Yes | No |
| gate/up | prefill/unified | 1 | No | Yes | Yes | No |
| QKV | decode | 1 | Yes | Yes | Yes | Yes |
| gate/up | decode | 1 | No | Yes | Yes | No |
| QKV | prefill/unified | 2 | No | No | No | No |
| gate/up | prefill/unified | 2 | No | No | No | No |
| QKV | decode | 2 | No | No | No | Yes |
| gate/up | decode | 2 | No | No | No | Yes |

### TP1 decode-QKV absolute measurements

| Concurrency | Repeat | Baseline TPOT | Fused TPOT | Baseline throughput | Fused throughput |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | 1 | 11.485 ms | 11.188 ms | 87.14 | 89.35 |
| 1 | 2 | 11.550 ms | 11.227 ms | 86.29 | 88.82 |
| 8 | 1 | 13.458 ms | 13.274 ms | 74.17 | 75.24 |
| 8 | 2 | 13.488 ms | 13.149 ms | 74.04 | 75.96 |

Its aggregate performance fields are:

```text
mean latency improvement = 2.3175%
direction consistent     = true
no large regression      = true
throughput consistent    = true
```

### Hardware scope

These measurements were collected on SM86 only.

Because cuBLASLt algorithm selection and BF16 reduction order may change with
GPU architecture and CUDA/cuBLAS version, this report does not assume that the
same numerical or performance behavior applies to other GPU architectures.

These results are provided as the current SM86 validation record. Additional
measurements on other GPU architectures can be compared against the same matrix
before making a broader production-policy decision.

**Comment by xiaguan** [comment]
Thanks for the thorough validation report. Please go ahead and open the PR.
